<a href="https://colab.research.google.com/github/lana994/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/Lana_Elgaddafi_LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

Stretch goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

In [1]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2019-11-04 22:14:41--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.1s    

2019-11-04 22:14:42 (127 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [0]:
import pandas as pd
import numpy as np
from scipy.stats import ttest_ind, ttest_ind_from_stats, ttest_rel

In [10]:
column_headers= ['party','handicapped-infants','water-project',
                          'budget','physician-fee-freeze', 'el-salvador-aid',
                          'religious-groups','anti-satellite-ban',
                          'aid-to-contras','mx-missile','immigration',
                          'synfuels', 'education', 'right-to-sue','crime','duty-free',
                          'south-africa']

df=pd.read_csv('house-votes-84.data', header=None, names=column_headers, na_values="?")
df.head()

party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n          NaN
2    democrat                 NaN             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [34]:
# to calculate the mean and t-test dist we need to change the values from "str" to "int"
df= df.replace({"y":1,"n":0})
df

party  handicapped-infants  ...  duty-free  south-africa
0    republican                  0.0  ...        0.0           1.0
1    republican                  0.0  ...        0.0           NaN
2      democrat                  NaN  ...        0.0           0.0
3      democrat                  0.0  ...        0.0           1.0
4      democrat                  1.0  ...        1.0           1.0
..          ...                  ...  ...        ...           ...
430  republican                  0.0  ...        0.0           1.0
431    democrat                  0.0  ...        0.0           1.0
432  republican                  0.0  ...        0.0           1.0
433  republican                  0.0  ...        0.0           1.0
434  republican                  0.0  ...        NaN           0.0

[435 rows x 17 columns]

In [35]:
df.isnull().sum()

party                     0
handicapped-infants      12
water-project            48
budget                   11
physician-fee-freeze     11
el-salvador-aid          15
religious-groups         11
anti-satellite-ban       14
aid-to-contras           15
mx-missile               22
immigration               7
synfuels                 21
education                31
right-to-sue             25
crime                    17
duty-free                28
south-africa            104
dtype: int64

In [36]:
df['party'].value_counts()

democrat      267
republican    168
Name: party, dtype: int64

In [37]:
rep=df[df['party']=='republican']
df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [38]:
dem=df[df['party']=='democrat']
df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

 ### PART 1: democrats support more than republican with p < 0.01

**Null hypothesis:**

there is no difference between the democratic and republican on the level of support for the the  "budget" bill

**Alternative hypotheresis:**
there is a difference between the democratic and republican on the level of support for the the  "budget" bill

**confidence level ** p <0.01

In [40]:
dem['budget'].mean()

0.8884615384615384

In [39]:
rep['budget'].mean()

0.13414634146341464

In [54]:
ttest_ind(dem['budget'],rep['budget'],nan_policy='omit')

Ttest_indResult(statistic=23.21277691701378, pvalue=2.0703402795404463e-77)

b/c the p- value <0.01 (2.07034e-77) we **reject** the **null hypothesis** and fail to reject the alternative hypothesis that there is a difference b/t republican and democratics on the support for 'budget' bill  

### PART 2 republicans support more than democrats with p < 0.01

**Null hypothesis:**
no diffrence b/t democratic & republican on the support for "religious-groups" bill   

**Alternative hypotheresis:**
 there is a diffrence b/t the republican & democratic on the support for "'religious-groups'"

** confidence level ** p <0.01

In [45]:
dem['religious-groups'].mean()

0.47674418604651164

In [46]:
rep['religious-groups'].mean()

0.8975903614457831

In [53]:
ttest_ind(dem['religious-groups'],rep['religious-groups'],nan_policy="omit")

Ttest_indResult(statistic=-9.737575825219457, pvalue=2.3936722520597287e-20)

b/c the p- value <0.01 (2.393e-20) we **reject** the **null hypothesis** and fail to reject the alternative hypothesis that there is a difference b/t republican and democratics on the support for ''religious-groups'' bill 

### PART 3: the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

**Null hypothesis:** no diffrence b/t democratic & republican on the support for "immigration" bill

**Alternative hypotheresis**: there is a diffrence b/t the republican & democratic on the support for "'immigration'"

 **confidence level** * p >0.1

In [55]:
df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [96]:
rep['immigration'].mean()

0.5575757575757576

In [97]:
dem['immigration'].mean()

0.4714828897338403

In [98]:
ttest_ind(rep['immigration'],dem['immigration'], nan_policy='omit')

Ttest_indResult(statistic=1.7359117329695164, pvalue=0.08330248490425066)

b/c the p- value is not > 0.1 (0.08330) we **fail to reject** the null hypothesis and reject the alternative hypothesis that there is a difference b/t republican and democratics on the support for 'immigration' bill 

## Stretch goals:

   #### Refactor your code into functions so it's easy to rerun with arbitrary variables
   #### Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)


In [104]:
ttest_ind(rep['water-project'],dem['water-project'],nan_policy='omit')

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

In [0]:
col = rep['water-project']
rep_water_project_no_nans = col[~np.isnan(col)]

col = dem['water-project']
dem_water_project_no_nans = col[~np.isnan(col)]





In [107]:
ttest_ind(rep_water_project_no_nans,dem_water_project_no_nans)

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823994811)

In [0]:
col=dem['budget']
col=rep['budget']
dem_budget_no_nans=col

col=dem['']
col=rep['']

col=dem['']
col=rep['']
